runtime/sema/check_casting_expression.go

```diff
@@ -154,7 +158,12 @@ func (checker *Checker) VisitCastingExpression(expression *ast.CastingExpression
 		return rightHandType
 
 	case ast.OperationCast:
-		if checker.expectedType != nil && checker.expectedType.Equal(rightHandType) {
+		// If there are errors in the lhs-expr, then the target type is considered as
+		// the inferred-type of the expression. i.e: exprActualType == rightHandType
+		// Then, it is not possible to determine whether the target type is redundant.
+		// Therefore don't check for redundant casts, if there are errors.
+		if !hasErrors &&
+			IsCastRedundant(leftHandExpression, exprActualType, rightHandType, checker.expectedType) {
```

Comment:
Thinking about hints and how the recent community effort to build a Cadence linter a bit more: I think it's great to have the hints, but maybe we should only have them during development, instead of always, i.e. also during production execution on the networks.

As a first step we could put code that generates hints, like this one, behind an interpreter option which is disabled by default. We would have it enabled it in the language server. 

This hint calculation looks a bit expensive, so maybe it would be good to gate it already in this PR.

Further, (and out of scope for this PR) we could maybe also factor the hints out of the type checker and have them be an optional second pass, or have them as linting/analyzer passes, like in Go: https://arslan.io/2019/06/13/using-go-analysis-to-write-a-custom-linter/.

Reply:
yeah, it would be really clean to separate out the linter to a separate phase, then we can also enable/disable the entire phase, rather than gating everywhere. It might need some additional effort though, so like you've suggested, would be good to handle it separately. I'll create an issue to track.

Will add a flag to enable/disable this only check for now.